### DISCUSSION TITLE
[Go] Why isnt this way faster?

### DISCUSSION THREAD

**Discussion by ghost** [PlummersSoftwareLLC/Primes#261] [Improvement suggestions]
Hey,

Did I make a mistake? Because it takes 300 milliseconds to complete.

```
package main

import (
	"fmt"
	"math"
	"sync"
	"time"
)

var (
	wg          sync.WaitGroup
	maxNumber   = 1000000
)

func main() {
	t0 := time.Now()
	for loop := 0; loop <= maxNumber; loop++ {
		wg.Add(1)
		go primeCheck(loop)
	}
	wg.Wait()
	t1 := time.Now()
	fmt.Printf("Time: %v \n", t1.Sub(t0))
}

func primeCheck(number int) {
	for i := 2; i <= int(math.Floor(float64(number)/2)); i++ {
		if number > 1 {
			break
		}
	}
	wg.Done()
}
```

**Comment by rbergen** [comment]
It may be helpful for your (potential) audience if you add the programming language this discussion concerns, to this topic's title :)

**Reply by ghost** [reply]
Hey,
Out of everything, which was the fastest?

I'd want to study rust instead of C | C++, but I'm not sure if I should or if the language is doomed from the start.

**Reply by rbergen** [reply]
Well, three things cross my mind:

1. This is a separate discussion, unrelated to the topic you opened above.
2. There is no final ranking yet.
3. I think Rust being "doomed" or not is not at all related to its performance in this drag-race. Looking at what Rust is now already used for (including Linux kernel development), my bet is that it is here to stay.

**Comment by Davo36** [comment]
So you're creating 1 million go routines? And then doing nothing much with them and closing them, then outputting the time right?

It seems to me that that's quite a lot to do in 0.3 secs isn't it?

**Comment by Davo36** [comment]
Keep going mate, it will be interesting to see how fast Go can be.  It would be cool if it can get close to some of the C and C++ versions.

**Reply by kpym** [reply]
With my [multithread solution](https://github.com/PlummersSoftwareLLC/Primes/pull/450) I obtain more then 21000 passes in 5s on my Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz.

But I have no idea how this compares with the fastest C/C++/Rust solutions.

**Reply by rbergen** [reply]
Until we have our solution in place with which people can share their benchmark results for others to look at, the best way to get an answer to that question is running a benchmark on the full set of solutions, on your own systems.

**Reply by kpym** [reply]
Can't github actions run all the benchmarks and export the results as `csv` in `Releases` for example ?

**Reply by ghost** [reply]
Yes,
Also why `csv`, thats trash. `json` is better.

**Reply by marghidanu** [reply]
> Can't github actions run all the benchmarks and export the results as `csv` in `Releases` for example ?

We already discussed this on several occasions. GitHub Actions is not suitable for this for several reasons. Initially, we had this in place and took it out. We will have machines that run the benchmark suite regularly and export the results soon.

**Reply by marghidanu** [reply]
> Yes,
> Also why `csv`, thats trash. `json` is better.

The benchmark tool offers both; feel free to use whatever you want. But CSV results and be imported directly in Excel (or other similar tools), I can't say as much for JSON.
